mjpc/tasks/humanoid/tracking/task.cc

```diff
@@ -0,0 +1,229 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "tasks/humanoid/tracking/task.h"
+
+#include <algorithm>
+#include <array>
+#include <cassert>
+#include <cmath>
+#include <cstring>
+#include <iostream>
+#include <map>
+
+#include <mujoco/mujoco.h>
+#include "../../../task.h"
+#include "utilities.h"
+
+
+static std::tuple<int, int, double, double> compute_interpolation_values(
+  double index, int max_index) {
+  int index_0 = std::floor(std::clamp(index, 0.0, (double)max_index));
+  int index_1 = std::min(index_0 + 1, max_index);
+
+  double weight_1 = std::clamp(index, 0.0, (double)max_index)
+    - index_0;
+  double weight_0 = 1.0 - weight_1;
+
+  return {index_0, index_1, weight_0, weight_1};
+}
+
+
+namespace mjpc {
+
+// ------------- Residuals for humanoid tracking task -------------
+//   Number of residuals: 
+//     Residual (0): Joint vel: minimise joint velocity
+//     Residual (1): Control: minimise control
+//     Residual (2-11): Tracking position: minimise tracking position error
+//         for {root, head, toe, heel, knee, hand, elbow, shoulder, hip}.
+//     Residual (11-20): Tracking velocity: minimise tracking velocity error
+//         for {root, head, toe, heel, knee, hand, elbow, shoulder, hip}.
+//   Number of parameters: 0
+// ----------------------------------------------------------------
+void humanoid::Tracking::Residual(const double* parameters,
+                                  const mjModel* model, const mjData* data,
+                                  double* residual) {
+  // ----- get mocap frames ----- //
+  // Hardcoded constant matching keyframes from CMU mocap dataset.
+  float fps = 30.0;
+  double current_index = data->time * fps;
+  int last_key_index = (model->nkey) - 1;
+
+  // Positions:
+  // We interpolate linearly between two consecutive key frames in order to
+  // provide smoother signal for tracking.
+  int key_index_0, key_index_1;
+  double weight_0, weight_1;
+  std::tie(key_index_0, key_index_1, weight_0, weight_1) =
+    compute_interpolation_values(current_index, last_key_index);
+
+  // ----- residual ----- //
+  int counter = 0;
+
+  // ----- joint velocity ----- //
+  mju_copy(residual + counter, data->qvel + 6, model->nv - 6);
+  counter += model->nv - 6;
+
+  // ----- action ----- //
+  mju_copy(&residual[counter], data->ctrl, model->nu);
+  counter += model->nu;
+
+  std::array<std::string, 16> body_names = {
+    "pelvis", "head", "ltoe", "rtoe", "lheel", "rheel", "lknee", "rknee",
+    "lhand", "rhand", "lelbow", "relbow", "lshoulder", "rshoulder", "lhip",
+    "rhip",
+  };
+
+  // Compute interpolated frame.
+  auto get_body_mpos = [=](std::string body_name, double result[3]) {
```

Comment:
Any reason to use `[=]`, rather than a by-reference default capture (`[&]`)? See [docs](https://en.cppreference.com/w/cpp/language/lambda#Lambda_capture).

Make body_name a `std::string_view` or `const std::string&`.

Same below with the other lambda.

Reply:
No good reason for not using by-reference capture. I changed this to explicitly use `[&]`.